labelme/widgets/canvas.py

```diff
@@ -155,6 +155,78 @@ def unHighlight(self):
     def selectedVertex(self):
         return self.hVertex is not None
 
+    def addLinePoint(self, pos):
+        self.current.addPoint(self.line[-1])
+
+        degree_increment = max(len(self.current.points) - self.current.points_in_segments - 1, 0)
+        self.line.createSegment(self.line[-1], degree_increment)
+
+        self.line.segments[0][1] += 1
+        ind = len(self.line.points) - 1
+        self.line.insertPoint(ind, pos)
+        self.line.points.append(pos)
+
+        if self.current.isClosed():
+            self.finalise()
+
+    def updateLine(self, pos):
+        if len(self.line) >= 3:
+            self.current.addPoint(self.line[-1], True, self.line[-3])
+        else:
+            self.current.addPoint(self.line[-1], True)
+        self.line.segments = []
+        degree_increment = max(len(self.current.points) - self.current.points_in_segments - 1, 0)
+        size = 2 + degree_increment
+        ind = max(len(self.current.points) - size, 0)
+        self.line.points = self.current[ind:]
+        self.line.points.append(pos)
+        if self.current.isClosed():
+            self.finalise()
+
+    # fix coordinates if summetric point is out of Pixmap
+    def updatePixmapPoints(self, x, y, p1, p2):
+        max_width = self.pixmap.size().width() - 1
+        max_height = self.pixmap.size().height() - 1
+        xa = p1.x()
+        xb = p2.x()
+        ya = p1.y()
+        yb = p2.y()
+        edgeX, edgeY = x, y
+
+        if xa - xb != 0:
+            edgeX = min(max_width, max(edgeX, 0))
+            edgeY = (ya - yb) / (xa - xb) * edgeX + (- xb * ya + yb * xa) / (xa - xb)
```

Comment:
please define operations you are using like coefficients. Now it is code duplication here

Reply:
done